File: TOOLKIT/hdfeos/samples/writedimscaleswath.f
	
!/*
! * In this example we will (1) open the "SwathFile" HDF file, (2) attach to
! * the "swid" swath, and (3) set dimension scale for a dimension in a filed.
! */

	program writedimscaleswath

	integer       j, status, swdetach, swclose
	integer*4     swfid, swid, swopen, swattach
	integer*4     swsetdimscale, swsetdimstrs
	real*8        BuffGeoTrack(20)
	real*4        BuffGeoTrack2(10)
	integer*4     GTdim,GTdim1,GTdim2,GTdim3
	integer*4     bands(15)
	integer*4     IBuffGeoTrack(40)
	integer*4     IBuffGeoTrack2(20)
	integer DFACC_RDWR
	parameter (DFACC_RDWR=3)
	integer DFNT_FLOAT32
	parameter (DFNT_FLOAT32=5)
	integer DFNT_FLOAT64
	parameter (DFNT_FLOAT64=6)
	integer DFNT_INT32
	parameter (DFNT_INT32=24)

!    /*
!     * We first open the HDF swath file, "SwathFile.hdf".  Because this file
!     * already exist and we wish to write to it, we use the DFACC_RDWR access
!     * code in the open statement.  The swopen routine returns the swath file
!     * id, swfid, which is used to identify the file in subsequent routines.
!     */

	swfid = swopen("SwathFile_created_with_hadeos_sample_file_write"//
     1"r_of_HDFEOS2_version_219_or_higher_release.hdf", DFACC_RDWR)

!    /*
!     * If the swath file cannot be found, swopen will return -1 for the file
!     * handle (swfid).  We there check that this is not the case before
!     * proceeding with the other routines.
!     * 
!     * The swattach routine returns the handle to the existing swath "swid",
!     * swid.  If the swath is not found, swattach returns -1 for the handle.
!     */
	      write(*,*) '=================================='
	if (swfid .ne. -1) then

!	/* the field Density has GeoTrack dimensions.
!	   the field Spectra has Res2xtr,Res2tr,Bands dimensions.
!	   see setupswath.f and defineflds.f
!	   nbands = 15;
!	   GeoTrack = 20;
!	*/
	  swid = swattach(swfid, "Swath1")
	  
	  if (swid .ne. -1) then
	    GTdim = 20
	    GTdim1 = 40
	    GTdim2 = 20
	    GTdim3 = 10
	    
	    do j=1,GTdim1
	       IBuffGeoTrack(j) = j
	    enddo
	    do j=1,GTdim
	       BuffGeoTrack(j) = j*2.5
	    enddo
	    status = swsetdimscale(swid,"Pressure","Res2tr",GTdim1,
     1	    DFNT_INT32, IBuffGeoTrack)
	    write(*,*) 'status swsetdimscale:Pressure:Res2tr=',status
	    status = swsetdimstrs(swid, "Pressure", "Res2tr", 
     1	    "Res2tr", "degrees", "I2")
	     
	     
	    status = swsetdimscale(swid,"Pressure","Res2xtr",GTdim2,
     1	    DFNT_INT32, IBuffGeoTrack2)
	    write(*,*) 'status swsetdimscale:Pressure:Res2xtr=',status

	    status = swsetdimscale(swid,"Density","GeoTrack",GTdim,
     1	    DFNT_FLOAT64, BuffGeoTrack)
	    write(*,*) 'status swsetdimscale:Density=',status,
     1      '    Error expected. Field is 1-dimensional'
	    
	    status = swsetdimscale(swid,"Temperature","GeoTrack",GTdim,
     1	    DFNT_FLOAT64, BuffGeoTrack)
	    write(*,*) 'status swsetdimscale:Temperature:GeoTrack=',
     1	    status
	    status = swsetdimstrs(swid, "Temperature", "GeoTrack", 
     1	    "GeoTrack", "degrees", "F7.2")
	    write(*,*) 'status swsetdimstrs:Temperature:GeoTrack=',
     1      status
	    
	    status = swsetdimscale(swid, "Temperature", "GeoXtrack",
     1	    GTdim3,DFNT_FLAOT32, BuffGeoTrack2)
	    write(*,*) 'status swsetdimscale:Temperature:GeoXtrack=',
     1	    status
	    
	    nbands = 15;
	    do j=1,nbands
	       bands(j) = j*2+1
	    enddo
	    status = swsetdimscale(swid, "Spectra", "Bands", nbands, 
     1	    DFNT_INT32, bands);
	    write(*,*) 'status swsetdimscale:Spectra =',status
	    
	    status = swsetdimstrs(swid, "Spectra", "Bands", "BANDS", 
     1	    "None", "I2")
	    
	    status = swdetach(swid)
	 endif
	 status = swclose(swfid)
	endif
	
	stop
	end
	

File: TOOLKIT/hdfeos/samples/WriteDimscaleSwath.c
#include "hdf.h"
#include "HdfEosDef.h"

/*
 * In this example we will (1) open the "SwathFile" HDF file, (2) attach to
 * the "Swath1" swath, and (3) set dimension scale for a dimension in a filed.
 */


main()
{

    intn            status, i, j;
    int32           swfid, SWid;
    int32           nbands, nGeoTrack;
    int32           bands[15]={3,6,9,12,15,18,23,26,29,32,33,34,36,37,39};
    char            label[16];
    char            unit[16];
    char            format[16];
    float           dataGeoTrack[20]={1.,2.,3.,6.,9.,12.,14.,15.,16.,17.,18.,23.,26.,29.,32.,33.,34.,36.,37.,39.};

    /*
     * We first open the HDF swath file, "SwathFile.hdf".  Because this file
     * already exist and we wish to write to it, we use the DFACC_RDWR access
     * code in the open statement.  The SWopen routine returns the swath file
     * id, swfid, which is used to identify the file in subsequent routines.
     */

    swfid = SWopen("SwathFile_created_with_hadeos_sample_file_writer_of_HDFEOS2_version_219_or_higher_release.hdf", DFACC_RDWR);
    if (swfid == -1)
      {
	printf("\t\tError: Cannot open file \"SwathFile.hdf\"\n");
	return -1;
      }

    /*
     * If the swath file cannot be found, SWopen will return -1 for the file
     * handle (swfid).  We there check that this is not the case before
     * proceeding with the other routines.
     * 
     * The SWattach routine returns the handle to the existing swath "Swath1",
     * GDid.  If the swath is not found, SWattach returns -1 for the handle.
     */

    if (swfid != -1)
      {
	/* the field Spectra has Bands, GeoTrack, GeoXtrack dimensions. 
	   see SetupSwath.c and Defineflds.c
	   GeoTrack = 20;
	   GeoXtrack = 10;
	*/
	
	SWid = SWattach(swfid, "Swath1");
	if (SWid == -1)
	  {
	    printf("\t\tError: Cannot attach to swath \"Swath1\"\n");
	    SWclose(swfid);
	    return -1;
	  }


	nGeoTrack = 20;
	strcpy(label, "GeoTrack");
	strcpy(unit, "Degree");
	strcpy(format, "F2");

	status = SWsetdimscale(SWid, "Temperature", "GeoTrack", nGeoTrack, DFNT_FLOAT32, dataGeoTrack);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale for GeoTrack dimemnsion in field \"Temperature\"\n");
	    SWdetach(SWid);
	    SWclose(swfid);
	    return -1;
	  }
	status = SWsetdimstrs(SWid, "Temperature", "GeoTrack", label, unit, format);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale strs for GeoTrack dimemnsion in field \"Temperature\"\n");
	    SWdetach(SWid);
	    SWclose(swfid);
	    return -1;
	  }


	nbands =  15;
	
	status = SWsetdimscale(SWid, "Spectra", "Bands", nbands, DFNT_INT32, bands);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale for Bands dimemnsion in field \"Spectra\"\n");
	    SWdetach(SWid);
	    SWclose(swfid);
	    return -1;
	  }
	
	strcpy(label, "Bands");
	strcpy(unit, "none");
	strcpy(format, "I2");
	
	status = SWsetdimstrs(SWid, "Spectra", "Bands", label, unit, format);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale strs for Bands dimemnsion in field \"Spectra\"\n");
	    SWdetach(SWid);
	    SWclose(swfid);
	    return -1;
	  }
	
	SWdetach(SWid);
	
	SWclose(swfid);
	
	return 0;

      }
}
